packages/react/src/auto/shadcn/inputs/relationships/ShadcnAutoHasOneForm.tsx

```diff
@@ -0,0 +1,274 @@
+  DialogDescription,
+  DialogTrigger,
+}: Pick<
+  ShadcnElements,
+  | "Badge"
+  | "Button"
+  | "Command"
+  | "CommandItem"
+  | "CommandInput"
+  | "CommandLoading"
+  | "Label"
+  | "Input"
+  | "CommandList"
+  | "CommandEmpty"
+  | "CommandGroup"
+  | "Checkbox"
+  | "DropdownMenu"
+  | "DropdownMenuContent"
+  | "DropdownMenuItem"
+  | "DropdownMenuLabel"
+  | "DropdownMenuSeparator"
+  | "DropdownMenuTrigger"
+  | "ScrollArea"
+  | "Popover"
+  | "PopoverContent"
+  | "PopoverTrigger"
+  | "Dialog"
+  | "DialogContent"
+  | "DialogTrigger"
+  | "DialogHeader"
+  | "DialogTitle"
+  | "DialogDescription"
+  | "DialogFooter"
+  | "DialogClose"
+>) => {
+  const renderOptionLabel = makeShadcnRenderOptionLabel({ Label, Badge, Button });
+
+  const ShadcnComboInput = makeShadcnAutoComboInput({
+    Command,
+    CommandInput,
+    Label,
+    CommandItem,
+    CommandList,
+    CommandEmpty,
+    CommandLoading,
+    CommandGroup,
+    Checkbox,
+    ScrollArea,
+    Popover,
+    PopoverContent,
+    PopoverTrigger,
+    Input,
+  });
+
+  function ShadcnHasOneForm(props: {
+    field: string;
+    children: React.ReactNode;
+    renderSelectedRecord?: (record: Record<string, any>) => React.ReactNode;
+    primaryLabel?: OptionLabel;
+    secondaryLabel?: OptionLabel;
+    tertiaryLabel?: OptionLabel;
+  }) {
+    const { field } = props;
+    const { path, metadata } = useAutoRelationship({ field });
+    const {
+      setValue,
+      getValues,
+      formState: { defaultValues, submitCount, isSubmitSuccessful },
+    } = useFormContext();
+
+    const { record, relatedModelOptions } = useHasOneController(props);
+    const [actionsOpen, setActionsOpen] = useState(false);
+    const [searchOpen, setSearchOpen] = useState(false);
+    const [modalOpen, setModalOpen] = useState(false);
+
+    const {
+      search,
+      searchFilterOptions,
+      pagination,
+      relatedModel: { records, fetching: isLoading },
+    } = relatedModelOptions;
+    const relationshipContext = useRelationshipContext();
+    const pathPrefix = relationshipContext?.transformPath ? relationshipContext.transformPath(props.field) : props.field;
+
+    const defaultRecordId = get(defaultValues, path)?.id;
+
+    // each time the form is submitted if the child record is created we need to set the id to the default record id
+    // that comes from the response to the action mutation
+    useEffect(() => {
+      if (isSubmitSuccessful && record && !record.id && !("_link" in record) && !("_unlink" in record) && defaultRecordId) {
+        setValue(path + ".id", defaultRecordId);
+      }
+    }, [record, defaultRecordId, path, setValue, submitCount, isSubmitSuccessful]);
+
+    const primaryLabel = useOptionLabelForField(field, props.primaryLabel);
+    const hasRecord = !!(record && !("_unlink" in record && record._unlink));
+    const recordOption = record ? getRecordAsOption(record, primaryLabel, props.secondaryLabel, props.tertiaryLabel) : null;
```

Comment:
Think we can push some of this logic into the controller or a new utility function? It all seems pretty smart such that duplicating it between the two implementations seems like asking for bugs in one but not the other

Reply:
okay, working on it.